components/supervisor-api/notification.proto

```diff
@@ -11,60 +11,126 @@ import "google/api/annotations.proto";
 option go_package = "github.com/gitpod-io/gitpod/supervisor/api";
 option java_package = "io.gitpod.supervisor.api";
 
-// Notification serivce allows external processes to notify the user and ask for decisions.
+// Notification serivce allows external processes to notify the user and ask for
+// decisions.
 service NotificationService {
 
-    // Prompts the user and asks for a decision. Typically called by some external process.
-    // If the list of actions is empty this service returns immediately,
-    // otherwise it blocks until the user has made their choice.
-    rpc Notify(NotifyRequest) returns (NotifyResponse) {
-        option (google.api.http) = {
-            post: "/v1/notification/notify"
-        };
-    }
-
-    // Subscribe to notifications. Typically called by the IDE.
-    rpc Subscribe(SubscribeRequest) returns (stream SubscribeResponse) {
-        option (google.api.http) = {
-            get: "/v1/notification/subscribe"
-        };
-    }
-
-    // Report a user's choice as a response to a notification. Typically called by the IDE.
-    rpc Respond(RespondRequest) returns (RespondResponse) {
-        option (google.api.http) = {
-            post: "/v1/notification/respond"
-        };
-    }
+  // Prompts the user and asks for a decision. Typically called by some external
+  // process. If the list of actions is empty this service returns immediately,
+  // otherwise it blocks until the user has made their choice.
+  rpc Notify(NotifyRequest) returns (NotifyResponse) {
+    option (google.api.http) = {
+      post : "/v1/notification/notify"
+    };
+  }
+
+  // Subscribe to notifications. Typically called by the IDE.
+  rpc Subscribe(SubscribeRequest) returns (stream SubscribeResponse) {
+    option (google.api.http) = {
+      get : "/v1/notification/subscribe"
+    };
+  }
+
+  // Report a user's choice as a response to a notification. Typically called by
+  // the IDE.
+  rpc Respond(RespondRequest) returns (RespondResponse) {
+    option (google.api.http) = {
+      post : "/v1/notification/respond"
+    };
+  }
+
+  // SubscribeActive to inform supervisor about
+  // which is the latest client actively used by the user.
+  // Typically called by the IDE.
+  // Will response what client should do like `open` `preview`
+  rpc SubscribeActive(SubscribeActiveRequest)
```

Comment:
Please clarify contract:
- Active means that a client is last focused.
- If a client should not respond after loosing focus then it should close the stream.
- There can be only one active client. The last one wins. For previous one the stream gets closed.

Reply:
good shout, I've extended the comment clarifying the behaviour 👌